lib/ObjWriter/objwriter.cpp

```diff
@@ -103,207 +103,221 @@ bool ObjectWriter::Init(llvm::StringRef ObjectFilePath) {
     return error("no subtarget info for target " + TripleName);
 
-  MCE = TheTarget->createMCCodeEmitter(*MII, *MRI, *MC);
-  if (!MCE)
+  CodeEmmiter =
+      TheTarget->createMCCodeEmitter(*InstrInfo, *RegisterInfo, *OutContext);
+  if (!CodeEmmiter)
     return error("no code emitter for target " + TripleName);
 
-  MAB = TheTarget->createMCAsmBackend(*MRI, TripleName, MCPU, MCOptions);
-  if (!MAB)
+  AsmBackend = TheTarget->createMCAsmBackend(*RegisterInfo, TripleName, MCPU,
+                                             TargetMOptions);
+  if (!AsmBackend)
     return error("no asm backend for target " + TripleName);
 
-  MS = TheTarget->createMCObjectStreamer(TheTriple, *MC, *MAB, *OS, MCE, *MSTI,
-                                         RelaxAll,
-                                         /*IncrementalLinkerCompatible*/ true,
-                                         /*DWARFMustBeAtTheEnd*/ false);
-  if (!MS)
+  Streamer = (MCObjectStreamer *)TheTarget->createMCObjectStreamer(
+      TheTriple, *OutContext, *AsmBackend, *OS, CodeEmmiter, *SubtargetInfo,
+      RelaxAll,
+      /*IncrementalLinkerCompatible*/ true,
+      /*DWARFMustBeAtTheEnd*/ false);
+  if (!Streamer)
     return error("no object streamer for target " + TripleName);
+  Assembler = &Streamer->getAssembler();
 
-  TM.reset(TheTarget->createTargetMachine(TripleName, MCPU, FeaturesStr,
-                                          TargetOptions(), None));
-  if (!TM)
+  TMachine.reset(TheTarget->createTargetMachine(
+      TripleName, MCPU, FeaturesStr, TargetOptions(), None));
+  if (!TMachine)
     return error("no target machine for target " + TripleName);
 
-  Asm.reset(TheTarget->createAsmPrinter(*TM, std::unique_ptr<MCStreamer>(MS)));
-  if (!Asm)
+  AssemblerPrinter.reset(TheTarget->createAsmPrinter(
+      *TMachine, std::unique_ptr<MCStreamer>(Streamer)));
+  if (!AssemblerPrinter)
     return error("no asm printer for target " + TripleName);
 
   FrameOpened = false;
   FuncId = 1;
 
+  SetCodeSectionAttribute("text", CustomSectionAttributes_Executable, nullptr);
+
   return true;
 }
 
-void ObjectWriter::Finish() { MS->Finish(); }
+void ObjectWriter::Finish() { Streamer->Finish(); }
 
 void ObjectWriter::SwitchSection(const char *SectionName,
                                  CustomSectionAttributes attributes,
                                  const char *ComdatName) {
-  auto &OST = *Asm->OutStreamer;
-  MCContext &OutContext = OST.getContext();
-  const MCObjectFileInfo *MOFI = OutContext.getObjectFileInfo();
-  Triple TheTriple(TripleName);
+  MCSection *Section = GetSection(SectionName, attributes, ComdatName);
+  Streamer->SwitchSection(Section);
+  if (Sections.count(Section) == 0) {
+    Sections.insert(Section);
+    if (!Section->getBeginSymbol()) { // Is it set for default sections?
+      MCSymbol *SectionStartSym = OutContext->createTempSymbol();
+      Streamer->EmitLabel(SectionStartSym);
+      Section->setBeginSymbol(SectionStartSym);
```

Comment:
We need this symbols and they are used in dwarf for EmitLabelPlusOffset, where label is our beginSymbol.